File: Greedy/codeforces_2045C.py

# Problem link ===>> https://codeforces.com/problemset/problem/2045/C


def main():
    S, T = list(map(str, input()))
    n, m = len(S), len(T)

    # Dictionary to store the maximum right-most occurrence of characters in T
    char_map = {}
    for i in range(m - 2, -1, -1):
        char_map[T[i]] = max(char_map.get(T[i], 0), i)

    min_length = float('inf')
    result = "-1"

    # Check for the minimum length concatenated substring
    for i in range(1, n):
        if S[i] not in char_map:
            continue

        current_length = i + m - char_map[S[i]]
        if current_length < min_length:
            min_length = current_length
            prefix = S[:i]
            suffix = T[char_map[S[i]]:]
            result = prefix + suffix

    return result
        


if __name__ == "__main__":
    main()

File: Greedy/codeforces_2045C.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/2045/C
// submission link ===>>

#include <bits/stdc++.h>
#define ll long long
using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    string S, T;
    cin >> S >> T;

    int n = S.size(), m = T.size();

    // Map to store the maximum right-most occurrence of characters in T
    map<char, int> mp;
    for (int i = m - 2; i >= 0; --i)
    {
        mp[T[i]] = max(mp[T[i]], i);
    }

    int minLength = INT_MAX;
    string result = "-1";

    // Check for the minimum length concatenated substring
    for (int i = 1; i < n; ++i)
    {
        if (mp.find(S[i]) == mp.end())
            continue;

        int currentLength = i + m - mp[S[i]];
        if (currentLength < minLength)
        {
            minLength = currentLength;
            string prefix = S.substr(0, i);
            string suffix = T.substr(mp[S[i]]);
            result = prefix + suffix;
        }
    }

    cout << result << '\n';
    return 0;
}
